Source/Core/VideoCommon/VertexLoaderX64.cpp

```diff
@@ -0,0 +1,453 @@
+	native_format->enable = true;
+	native_format->offset = m_dst_ofs;
+	native_format->type = VAR_FLOAT;
+	native_format->integer = false;
+	m_dst_ofs += store_bytes;
+	return load_bytes;
+}
+
+void VertexLoaderX64::ReadColor(OpArg data, u64 attribute, int format, int elements)
+{
+	int load_bytes = 0;
+	static const u32 mask_565 = 0xF8FCF800;
+	static const u32 mask_0f  = 0x0F0F0F0F;
+	static const u32 mask_f0  = 0xF0F0F0F0;
+	static const u32 mask_fc  = 0xFCFCFCFC;
+	switch (format)
+	{
+		case FORMAT_24B_888:
+		case FORMAT_32B_888x:
+		case FORMAT_32B_8888:
+			MOV(32, R(scratch1), data);
+			if (format != FORMAT_32B_8888 || !elements)
+				OR(32, R(scratch1), Imm32(0xFF000000));
+			MOV(32, MDisp(dst_reg, m_dst_ofs), R(scratch1));
+			load_bytes = 3 + (format != FORMAT_24B_888);
+			break;
+
+		case FORMAT_16B_565:
+			//                   RRRRRGGG GGGBBBBB
+			// AAAAAAAA BBBBBBBB GGGGGGGG RRRRRRRR
+			LoadAndSwap(16, scratch1, data);
+			if (cpu_info.bBMI2)
+			{
+				PDEP(32, scratch1, scratch1, M(&mask_565));
+				MOV(32, R(scratch2), R(scratch1));
+			}
+			else
+			{
+				MOV(32, R(scratch3), R(scratch1));
+				SHL(32, R(scratch1), Imm8(16));
+				AND(32, R(scratch1), Imm32(0xF8000000));
+
+				MOV(32, R(scratch2), R(scratch3));
+				SHL(32, R(scratch2), Imm8(13));
+				AND(32, R(scratch2), Imm32(0x00FC0000));
+				OR(32, R(scratch1), R(scratch2));
+
+				SHL(32, R(scratch3), Imm8(11));
+				AND(32, R(scratch3), Imm32(0x0000F800));
+				OR(32, R(scratch1), R(scratch3));
+				MOV(32, R(scratch2), R(scratch1));
+			}
+
+			SHR(32, R(scratch1), Imm8(5));
+			AND(32, R(scratch1), Imm32(0x07000700));
+			OR(32, R(scratch1), R(scratch2));
+
+			SHR(32, R(scratch2), Imm8(6));
+			AND(32, R(scratch2), Imm32(0x00030000));
+			OR(32, R(scratch1), R(scratch2));
+
+			OR(8, R(scratch1), Imm8(0xFF));
+			SwapAndStore(32, MDisp(dst_reg, m_dst_ofs), scratch1);
+			load_bytes = 2;
+			break;
+
+		case FORMAT_16B_4444:
+			//                   RRRRGGGG BBBBAAAA
+			// AAAAAAAA BBBBBBBB GGGGGGGG RRRRRRRR
+			LoadAndSwap(16, scratch1, data);
+			if (cpu_info.bBMI2)
+			{
+				PDEP(32, scratch2, scratch1, M(&mask_0f));
+				PDEP(32, scratch1, scratch1, M(&mask_f0));
+			}
+			else
+			{
+				MOV(32, R(scratch3), R(scratch1));
+				SHL(32, R(scratch1), Imm8(12));
+				AND(32, R(scratch1), Imm32(0x0F000000));
+				MOV(32, R(scratch2), R(scratch1));
+
+				MOV(32, R(scratch1), R(scratch3));
+				SHL(32, R(scratch1), Imm8(8));
+				AND(32, R(scratch1), Imm32(0x000F0000));
+				OR(32, R(scratch2), R(scratch1));
+
+				MOV(32, R(scratch1), R(scratch3));
+				SHL(32, R(scratch1), Imm8(4));
+				AND(32, R(scratch1), Imm32(0x00000F00));
+				OR(32, R(scratch2), R(scratch1));
+
+				AND(32, R(scratch3), Imm8(0x0F));
+				OR(32, R(scratch2), R(scratch3));
+
+				MOV(32, R(scratch1), R(scratch2));
+				SHL(32, R(scratch1), Imm8(4));
+			}
+			OR(32, R(scratch1), R(scratch2));
+			SwapAndStore(32, MDisp(dst_reg, m_dst_ofs), scratch1);
```

Comment:
isn't this swapping ond both loading and saving + the swizzling a bit redundant?